### PR TITLE
Adds a ENS address selector to the Profile page

### PR DESCRIPTION
**Pull request by ramirotw**
# Summary
Close https://github.com/cowprotocol/cowswap/issues/190

This adds a new AddressSelector component to the Profile page with all the ENS addresses the user has. It uses the official ENS subgraph

### CODE REVIEW

**Comment by ramirotw** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by elena-zh** [comment]
Hey @ramirotw , could you please provide me the PR preview link please?

**Comment by ramirotw** [comment]
@elena-zh the build is finally fixed.

Btw, the CLA check fails because of the prettier bot that commited [Fix code style issues with Prettier](https://github.com/cowprotocol/cowswap/pull/367/commits/828b2e1b6c43e94bc9d5240b6d2742690a94cc83)

```json
{
  id: 'MDQ6VXNlcjU5MjgzODYy',
  databaseId: 59283862,
  login: 'lint-action'
}
```

File: package.json [hunk 1]
```diff
@@ -102,7 +102,7 @@
     "eslint-plugin-simple-import-sort": "^7.0.0",
     "ethers": "^5.4.6",
     "graphql": "^15.5.0",
-    "graphql-request": "^3.4.0",
+    "graphql-request": "^4.2.0",
```
Comment (ramirotw) [inline package.json:105]: I introduced this change to force the CI script to run the `postinstall` npm script which runs the `yarn graphql:generate` script which generates the graphql TS types in `src/state/data/generated.ts` that is needed to import the `EnsNamesQuery` type.

**Comment by elena-zh** [comment]
Hey @ramirotw !
Nice changes!
Unfortunately, I can't finish testing until the issue https://github.com/cowprotocol/explorer/issues/63 is fixed.

Some tiny notes from the UI side: 
1. Maybe we should not show an 'expandable' dropdown when a connected wallet does not have an ENS name for an account? WDYT?
disable when not available
2. Once connected to an account with ENS name, maybe we should show the an affiliate link with ENS name  by default instead of the address, as the account info shows ENS name? 
image
3. Can we align link nicer in a mobile view? 
allign better

Thanks!

**Comment by ramirotw** [comment]
> 3\. Can we align link nicer in a mobile view?
>     allign better

@alongoni any idea on how we could improve this on mobile?

**Comment by ramirotw** [comment]
> 1. Maybe we should not show an 'expandable' dropdown when a connected wallet does not have an ENS name for an account? WDYT?
>    disable when not available

done

> 2\. Once connected to an account with ENS name, maybe we should show the an affiliate link with ENS name  by default instead of the address, as the account info shows ENS name?
>     image

I have already done this https://github.com/cowprotocol/cowswap/pull/367/commits/0bc157968b8ba33462dbef1a223620ea5fd34376#diff-091719e9c8faa4dd8e97d8f607ee1f288ba02bb4f6be40589de8de2411e589baL36 but there were a few issues while switching accounts and maintaining state so to keep things simple I removed it. Maybe we can add that feature after this gets merged.

**Comment by elena-zh** [comment]
Hey @ramirotw , seems that the last commit has failed, as i do not see any changes relates to the case 1. 
As for the case 2, I have created a separate issue for this case #404 .
@alongoni , could you please take a look at the case 3?

**Comment by elena-zh** [comment]
> > Hey @ramirotw , seems that the last commit has failed, as i do not see any changes relates to the case 1.
> 
> @elena-zh aren't you seeing this?
> 
> image

@ramirotw , I see this.. However, I thought that the fix should disable the dropdown when a user does not have an ENS name, so I thought that there were issues with the last commit. 
By disabling the dropdown I mean this: 
image

**Comment by ramirotw** [comment]
> I still see the arrow that should expand the dropdown, so it seems that it is still enabled.

that is the disabled state of the dropdown, or you mean to remove entirely the dropdown as we do for Rinkeby/GC?

File: src/custom/pages/Profile/AddressSelector.tsx [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    // the selected address is a ens name, verify that resolves to the correct address
+    const verify = async () => {
+      const resolvedAddress = await library?.resolveName(selectedAddress)
+      if (resolvedAddress !== address) {
+        dispatch(updateAddress(address))
+      }
+    }
+
+    verify()
+  }, [selectedAddress, address, dispatch, library])
```
Comment (alfetopito) [inline src/custom/pages/Profile/AddressSelector.tsx:78]: These two `useEffect` seem to depend on almost the same deps. Does it make sense to merge them into a single one?

File: src/custom/pages/Profile/ens.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import { SupportedChainId } from 'constants/chains'
+import { ClientError, gql, GraphQLClient } from 'graphql-request'
+import { EnsNamesQuery } from 'state/data/generated'
+
+// List of supported subgraphs. Note that the app currently only support one active subgraph at a time
```
Comment (alfetopito) [inline src/custom/pages/Profile/ens.ts:5]: What does this mean? That we can't have more than 1 ENS subgraph or any kind of subgraph?

**Comment by alongoni** [comment]
> > 3. Can we align link nicer in a mobile view?
> > allign better
> 
> @alongoni any idea on how we could improve this on mobile?

Maybe we can use the shorte version of the address e.g: 0x3d2a...1b21

**Comment by elena-zh** [comment]
> > I still see the arrow that should expand the dropdown, so it seems that it is still enabled.
> 
> that is the disabled state of the dropdown, or you mean to remove entirely the dropdown as we do for Rinkeby/GC?

Yes, I would remove it at all when an account does not have a ENS name

**Comment by elena-zh** [comment]
> > > 3. Can we align link nicer in a mobile view?
> > >    allign better
> > 
> > 
> > @alongoni any idea on how we could improve this on mobile?
> 
> Maybe we can use a shorter version of the address e.g: 0x3d2a...1b21

that works for me!

File: src/custom/pages/Profile/ens.ts (hunk 3 above)
Comment (ramirotw) [reply src/custom/pages/Profile/ens.ts]: I removed this comment and also added the rinkeby subgraph that was missing

**Comment by elena-zh** [comment]
Hey @ramirotw , great changes!
I was also able to check that the trade was recorded to the correct referrer (my previous tests) despite of the fact that the issue  https://github.com/cowprotocol/explorer/issues/63 is still here. 

However, **sometimes** I face this issue: when I share a referral link with an ENS name, the app shows me 'invalid' banner for a couple of seconds. 
More often I can face it when I paste the link to a new browser/1st time opening the link. Then, the issue becomes not reproducible. 
See the video: https://watch.screencastify.com/v/qvteh2AKFxhzxxUwUAMz

Could you please take a look into this issue? 

Thanks!

**Comment by ramirotw** [comment]
> Hey @ramirotw , great changes! I was also able to check that the trade was recorded to the correct referrer (my previous tests) despite of the fact that the issue [cowprotocol/explorer#63](https://github.com/cowprotocol/explorer/issues/63) is still here.
> 
> However, **sometimes** I face this issue: when I share a referral link with an ENS name, the app shows me 'invalid' banner for a couple of seconds. More often I can face it when I paste the link to a new browser/1st time opening the link. Then, the issue becomes not reproducible. See the video: https://watch.screencastify.com/v/qvteh2AKFxhzxxUwUAMz
> 
> Could you please take a look into this issue?
> 
> Thanks!

hey @elena-zh, I fixed both issues. Please run a few affiliate workflows to be sure nothing broke because I did some changes to fix the first invalid state

File: src/custom/hooks/useParseReferralQueryParam.ts [hunk 4]
```diff
@@ -1,7 +1,8 @@
-import { useMemo } from 'react'
+import { useMemo, useState } from 'react'
 import useParsedQueryString from 'hooks/useParsedQueryString'
 import { REFERRAL_QUERY_PARAM } from 'hooks/useReferralLink'
 import useENS from 'hooks/useENS'
+import { isAddress } from '../utils'
```
Comment (alfetopito) [inline src/custom/hooks/useParseReferralQueryParam.ts:5]: Can you use the full path here? I think it should be

```suggestion
import { isAddress } from 'hook/utils'
```

**Review by elena-zh — APPROVED** [review]
Hey @ramirotw , changes look great!

The only issue I can mention is that due to Explorer issue https://github.com/cowprotocol/explorer/issues/63 Cowswap app does not receive a successful transaction pp-up, so a referral URL is not removed from the local storage after a successful trade and remains to be displayed. 
But this issue is not related to your changes, so the PR is approved. 

Thanks

File: src/custom/hooks/useParseReferralQueryParam.ts (hunk 4 above)
Comment (alfetopito) [reply src/custom/hooks/useParseReferralQueryParam.ts]: I guess this was not the right path... sorry!

Screen Shot 2022-04-18 at 11 08 44

Will need to figure out the correct one

Comment (ramirotw) [reply src/custom/hooks/useParseReferralQueryParam.ts]: no worries! just fixed it

**Pull request merged**